### PR TITLE
Picking appropriate derived type when serializing/deserializing

### PR DESCRIPTION
**Pull request by dsgouda**
## Description
ClientRuntime bug fix for issue under AutoRest https://github.com/Azure/autorest/issues/1842
When serializing/deserializing a derived type object which does not have a JsonObjectAttribute assigned to it, we end up picking its base type's JsonObjectAttribute and perform serialization/deserialization using that type. We should ignore the base type attribute and instead serialize using the derived type name itself.
Added test case 

@markcowl @fearthecowboy @olydis PTAL

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as a link to the swagger spec, used to generate the code.
- [ ] The `project.json` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@dsgouda__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (deshank). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

File: src/ClientRuntime/Microsoft.Rest.ClientRuntime/Serialization/PolymorphicSerializeJsonConverter.cs [hunk 1]
```diff
@@ -5,6 +5,7 @@
 using System.Linq;
 using System.Reflection;
 using Newtonsoft.Json;
+using System.IO;
```
Comment (olydis) [inline src/ClientRuntime/Microsoft.Rest.ClientRuntime/Serialization/PolymorphicSerializeJsonConverter.cs]: required?

File: src/ClientRuntime/Microsoft.Rest.ClientRuntime.Tests/JsonTransformationConverterTest.cs [hunk 2]
```diff
@@ -353,6 +354,28 @@ public void TestDeserializationOfResourceWithConflictingProperties()
 }";
             var newJson = JsonConvert.SerializeObject(deserializedResource, deserializeSettings);
             Assert.Equal(expectedSerializedJson, newJson);
-        }        
+        }
+
+        [Fact]
+        public void TestPolymorphicJsonDeserializer()
```
Comment (olydis) [inline src/ClientRuntime/Microsoft.Rest.ClientRuntime.Tests/JsonTransformationConverterTest.cs]: what about testing serialization? I assume there is an equivalently significant impact on that and you've touched `PolymorphicSerializeJsonConverter.cs` which I guess was covered by this test! ;-)

Comment (dsgouda) [inline src/ClientRuntime/Microsoft.Rest.ClientRuntime.Tests/JsonTransformationConverterTest.cs]: That's a fair point, test case is on its way.

**Comment by markcowl** [comment]
@azuresdkci retest this please

**Comment by cormacpayne** [comment]
@markcowl is this OK to merge?

**Review by markcowl — DISMISSED** [review]
Need to update the runtime version so we can release

**Pull request merged**